src/domain/common/canvas-checkout/canvas.checkout.lifecycle.options.provider.ts

```diff
@@ -0,0 +1,177 @@
+import { LifecycleService } from '@domain/common/lifecycle/lifecycle.service';
+import { EntityNotInitializedException } from '@common/exceptions';
+import { AgentInfo } from '@core/authentication';
+import { AuthorizationService } from '@core/authorization/authorization.service';
+import { IAuthorizationPolicy } from '@domain/common/authorization-policy';
+import { CanvasCheckoutEventInput } from './dto/canvas.checkout.dto.event';
+import { ICanvasCheckout } from './canvas.checkout.interface';
+import { CanvasCheckoutService } from './canvas.checkout.service';
+import { CanvasCheckoutStateEnum } from '@common/enums/canvas.checkout.status';
+
+@Injectable()
+export class CanvasCheckoutLifecycleOptionsProvider {
+  constructor(
+    private lifecycleService: LifecycleService,
+    private canvasCheckoutService: CanvasCheckoutService,
+    private authorizationService: AuthorizationService,
+    @Inject(WINSTON_MODULE_NEST_PROVIDER) private readonly logger: LoggerService
+  ) {}
+
+  async eventOnCanvasCheckout(
+    canvasCheckoutEventData: CanvasCheckoutEventInput,
+    agentInfo: AgentInfo
+  ): Promise<ICanvasCheckout> {
+    const canvasCheckout =
+      await this.canvasCheckoutService.getCanvasCheckoutOrFail(
+        canvasCheckoutEventData.canvasCheckoutID
+      );
+
+    if (!canvasCheckout.lifecycle)
+      throw new EntityNotInitializedException(
+        `Canvas checkout Lifecycle not initialized on CanvasCheckout: ${canvasCheckout.id}`,
+        LogContext.CONTEXT
+      );
+
+    // Send the event, translated if needed
+    this.logger.verbose?.(
+      `Event ${canvasCheckoutEventData.eventName} triggered on canvasCheckout: ${canvasCheckout.id} using lifecycle ${canvasCheckout.lifecycle.id}`,
+      LogContext.CONTEXT
+    );
+
+    await this.lifecycleService.event(
+      {
+        ID: canvasCheckout.lifecycle.id,
+        eventName: canvasCheckoutEventData.eventName,
+      },
+      this.CanvasCheckoutLifecycleMachineOptions,
+      agentInfo,
+      canvasCheckout.authorization
+    );
+
+    // Todo: needed to address a race condition
+    // Events that are triggered by XState are fire and forget. So they above event will potentially return
+    // before all the actions that were triggered by the event have completed.
+    // As in this case the events should all be quite fast a short wait is sufficient, but we need to fix this properly
+    await new Promise(resolve => setTimeout(resolve, 100));
+
+    return await this.canvasCheckoutService.getCanvasCheckoutOrFail(
+      canvasCheckout.id
+    );
+  }
+
+  private logMessage(msg: string) {
+    const logActions = true;
+    if (logActions) {
+      this.logger.verbose?.(
+        `[Lifecycle] Context checkout provider - ${msg}`,
+        LogContext.LIFECYCLE
+      );
+    }
+  }
+
+  private CanvasCheckoutLifecycleMachineOptions: Partial<
+    MachineOptions<any, any>
+  > = {
+    actions: {
+      availableEntry: (_, __) => {
```

Comment:
Weren't these log messages just to debug it?

Reply:
Yes, but really super useful so I put in a flag to be able to disable them while leaving them in. Also useful to be able to explain to others what is happening. 

I think we may find this kind of lifecycle "debugging" is critical for us to really be able to manage these lifecycles in the near-medium term as they are way to much of a black box at the moment.